ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiWorldStateArchive.java

```diff
@@ -0,0 +1,107 @@
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+package org.hyperledger.besu.ethereum.bonsai;
+
+import org.hyperledger.besu.ethereum.core.Address;
+import org.hyperledger.besu.ethereum.core.Hash;
+import org.hyperledger.besu.ethereum.core.MutableWorldState;
+import org.hyperledger.besu.ethereum.core.WorldState;
+import org.hyperledger.besu.ethereum.proof.WorldStateProof;
+import org.hyperledger.besu.ethereum.storage.StorageProvider;
+import org.hyperledger.besu.ethereum.storage.keyvalue.KeyValueSegmentIdentifier;
+import org.hyperledger.besu.ethereum.worldstate.WorldStateArchive;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import org.apache.tuweni.bytes.Bytes;
+import org.apache.tuweni.bytes.Bytes32;
+import org.apache.tuweni.units.bigints.UInt256;
+
+public class BonsaiWorldStateArchive implements WorldStateArchive {
+
+  private final BonsaiPersistedWorldState persistedState;
+  private final Map<Bytes32, BonsaiLayeredWorldState> layeredWorldStates;
+
+  public BonsaiWorldStateArchive(final StorageProvider provider) {
+    persistedState =
+        new BonsaiPersistedWorldState(
+            this,
+            provider.getStorageBySegmentIdentifier(KeyValueSegmentIdentifier.ACCOUNT_INFO_STATE),
+            provider.getStorageBySegmentIdentifier(KeyValueSegmentIdentifier.CODE_STORAGE),
+            provider.getStorageBySegmentIdentifier(
+                KeyValueSegmentIdentifier.ACCOUNT_STORAGE_STORAGE),
+            provider.getStorageBySegmentIdentifier(KeyValueSegmentIdentifier.TRIE_BRANCH_STORAGE),
+            provider.getStorageBySegmentIdentifier(KeyValueSegmentIdentifier.TRIE_LOG_STORAGE));
+    layeredWorldStates = new HashMap<>();
+  }
+
+  @Override
+  public Optional<WorldState> get(final Hash rootHash) {
+    if (layeredWorldStates.containsKey(rootHash)) {
+      return Optional.of(layeredWorldStates.get(rootHash));
+    } else if (rootHash.equals(persistedState.rootHash())) {
+      return Optional.of(persistedState);
+    } else {
+      return Optional.empty();
+    }
+  }
+
+  public void addLayeredWorldState(final BonsaiLayeredWorldState worldState) {
+    layeredWorldStates.put(worldState.rootHash(), worldState);
+  }
+
+  @Override
+  public boolean isWorldStateAvailable(final Hash rootHash) {
+    return layeredWorldStates.containsKey(rootHash)
+        || persistedState.rootHash().equals(rootHash) /* || check disk storage */;
+  }
+
+  @Override
+  public Optional<MutableWorldState> getMutable(final Hash rootHash) {
+    if (rootHash.equals(persistedState.rootHash())) {
+      return Optional.of(persistedState);
+    } else {
+      return Optional.empty();
+    }
+  }
+
+  @Override
+  public WorldState get() {
+    return null;
```

Comment:
`get(EMPTY_ROOT_HASH).get()` and https://github.com/hyperledger/besu/blob/7e52d4d85458ccf91bada647fdeb1ac363158f31/ethereum/core/src/main/java/org/hyperledger/besu/ethereum/worldstate/DefaultWorldStateArchive.java#L66

Reply:
org.hyperledger.besu.ethereum.worldstate.WorldStateArchive#get() is a dead method and should be removed.  No one calls it.  I'll remove it from the parent and the Forest world state too.